### PR TITLE
Case 20504: Trying to fix avatar entity issue

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/20504/AvatarEntity-with-finite-lifetime-vanishes-immediately

Test plan:
- Run a server from this PR.  Go there and have someone else join you.
- One of you runs this from the console:
```
Entities.addEntity({
    type: "Model",
    modelURL: "http://hifi-content.s3.amazonaws.com/alan/dev/Cowboy-hat.fbx?2",
    shapeType: "simple-hull",
    localPosition: { x:0, y:0.2, z:0 },
    name: "hat",
    dynamic: false,
    collisionless: true,
    parentID: MyAvatar.sessionUUID,
    parentJointIndex: MyAvatar.getJointIndex("Head"),
    lifetime: 20,
    dimensions: { x: 0.3239, y: 0.15, z: 0.3779 }
}, true);
```
- Both of you see the hat appear.  Both of you can select the hat in Create.  The hat disappears after 20 seconds.

(Andrew's edit: Note the **bug** on hifi/master that I was seeing was: the hat would vanish immediately.  It was being determined to have an expired `lifetime` because the `created` timestamp property was being overwritten to be 0 instead of the time at which it was actually created.)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16247/HighFidelity-Beta-PR14657-2c5789a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16247/HighFidelity-Beta-PR14657-2c5789a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16247/HighFidelity-Beta-Interface-PR14657-2c5789a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16247/HighFidelity-Beta-PR14657-2c5789a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16247/HighFidelity-Beta-Interface-PR14657-2c5789a.exe

File: libraries/entities/src/AnimationPropertyGroup.cpp [hunk 1]
```diff
@@ -309,14 +323,14 @@ EntityPropertyFlags AnimationPropertyGroup::getEntityProperties(EncodeBitstreamP
...
     requestedProperties += PROP_ANIMATION_FRAME_INDEX;
     requestedProperties += PROP_ANIMATION_PLAYING;
     requestedProperties += PROP_ANIMATION_LOOP;
     requestedProperties += PROP_ANIMATION_FIRST_FRAME;
     requestedProperties += PROP_ANIMATION_LAST_FRAME;
     requestedProperties += PROP_ANIMATION_HOLD;
-    requestedProperties += PROP_ANIMATION_ALLOW_TRANSLATION;
 
     return requestedProperties;
 }
```
Comment (SamGondelman) [inline libraries/entities/src/AnimationPropertyGroup.cpp]: Not the fix

File: libraries/entities/src/EntityItem.cpp [hunk 2]
```diff
@@ -1309,7 +1306,7 @@ EntityItemProperties EntityItem::getProperties(const EntityPropertyFlags& desire
     COPY_ENTITY_PROPERTY_TO_PROPERTIES(parentJointIndex, getParentJointIndex);
     COPY_ENTITY_PROPERTY_TO_PROPERTIES(queryAACube, getQueryAACube);
     COPY_ENTITY_PROPERTY_TO_PROPERTIES(canCastShadow, getCanCastShadow);
-    // COPY_ENTITY_PROPERTY_TO_PROPERTIES(isVisibleInSecondaryCamera, getIsVisibleInSecondaryCamera); // not sent over wire
+    COPY_ENTITY_PROPERTY_TO_PROPERTIES(isVisibleInSecondaryCamera, isVisibleInSecondaryCamera);
     withReadLock([&] {
         _grabProperties.getProperties(properties);
     });
```
Comment (SamGondelman) [inline libraries/entities/src/EntityItem.cpp]: Not the fix

File: libraries/entities/src/EntityItemProperties.cpp [hunk 3]
```diff
@@ -2679,16 +2679,15 @@ OctreeElement::AppendState EntityItemProperties::encodeEntityEditPacket(PacketTy
             APPEND_ENTITY_PROPERTY(PROP_DIMENSIONS, properties.getDimensions());
             APPEND_ENTITY_PROPERTY(PROP_ROTATION, properties.getRotation());
             APPEND_ENTITY_PROPERTY(PROP_REGISTRATION_POINT, properties.getRegistrationPoint());
-            // FIXME: deal with these
-            // APPEND_ENTITY_PROPERTY(PROP_CREATED, properties.getCreated());
-            // APPEND_ENTITY_PROPERTY(PROP_LAST_EDITED_BY, properties.getLastEditedBy());
-            // APPEND_ENTITY_PROPERTY(PROP_ENTITY_HOST_TYPE, (uint32_t)properties.getEntityHostType());
-            // APPEND_ENTITY_PROPERTY(PROP_OWNING_AVATAR_ID, properties.getOwningAvatarID());
+            APPEND_ENTITY_PROPERTY(PROP_CREATED, properties.getCreated());
+            APPEND_ENTITY_PROPERTY(PROP_LAST_EDITED_BY, properties.getLastEditedBy());
```
Comment (SamGondelman) [inline libraries/entities/src/EntityItemProperties.cpp]: ...possibly the fix?  (these props were requested but not appended/read)

File: libraries/entities/src/GrabPropertyGroup.cpp [hunk 4]
```diff
@@ -255,6 +262,7 @@ bool GrabPropertyGroup::setProperties(const EntityItemProperties& properties) {
     SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, GrabFollowsController, grabFollowsController, setGrabFollowsController);
     SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, Triggerable, triggerable, setTriggerable);
     SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, Equippable, equippable, setEquippable);
+    SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, GrabDelegateToParent, grabDelegateToParent, setGrabDelegateToParent);
     SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, EquippableLeftPosition, equippableLeftPosition, setEquippableLeftPosition);
     SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, EquippableLeftRotation, equippableLeftRotation, setEquippableLeftRotation);
     SET_ENTITY_GROUP_PROPERTY_FROM_PROPERTIES(Grab, EquippableRightPosition, equippableRightPosition,
```
Comment (SamGondelman) [inline libraries/entities/src/GrabPropertyGroup.cpp]: Not the fix (cc @sethalves)

File: libraries/entities/src/EntityItem.cpp [hunk 5]
```diff
@@ -85,15 +85,15 @@ EntityPropertyFlags EntityItem::getEntityProperties(EncodeBitstreamParams& param
     requestedProperties += PROP_DIMENSIONS;
     requestedProperties += PROP_ROTATION;
     requestedProperties += PROP_REGISTRATION_POINT;
-    // TODO: handle PROP_CREATED?
+    requestedProperties += PROP_CREATED;
```
Comment (SamGondelman) [inline libraries/entities/src/EntityItem.cpp]: Possibly the fix?  Things needed to be sending and receiving created so the lifetimes would work.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16259/HighFidelity-Beta-PR14657-1452d23.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16259/HighFidelity-Beta-PR14657-1452d23.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16259/HighFidelity-Beta-Interface-PR14657-1452d23.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16259/HighFidelity-Beta-PR14657-1452d23.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16259/HighFidelity-Beta-Interface-PR14657-1452d23.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16264/HighFidelity-Beta-PR14657-6e43ff6.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16264/HighFidelity-Beta-PR14657-6e43ff6.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16264/HighFidelity-Beta-Interface-PR14657-6e43ff6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16264/HighFidelity-Beta-PR14657-6e43ff6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16264/HighFidelity-Beta-Interface-PR14657-6e43ff6.exe

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp [hunk 6]
```diff
@@ -390,7 +390,7 @@ void Avatar::updateAvatarEntities() {
             QVariantMap asMap = variantProperties.toMap();
             QScriptValue scriptProperties = variantMapToScriptValue(asMap, scriptEngine);
             EntityItemProperties properties;
-            EntityItemPropertiesFromScriptValueHonorReadOnly(scriptProperties, properties);
+            EntityItemPropertiesFromScriptValueIgnoreReadOnly(scriptProperties, properties);
```
Comment (SamGondelman) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.cpp]: We want to read all properties from the script value, including read only values.

File: libraries/entities/src/EntityItem.cpp (hunk 5 above)
Comment (SamGondelman) [inline libraries/entities/src/EntityItem.cpp]: `created` is handled like any other property now.

File: libraries/entities/src/EntityItemProperties.cpp [hunk 7]
```diff
@@ -1367,11 +1363,6 @@ QScriptValue EntityItemProperties::copyToScriptValue(QScriptEngine* engine, bool
     if (!psuedoPropertyFlagsActive || psueudoPropertyFlags.test(EntityPsuedoPropertyFlag::Type)) {
         COPY_PROPERTY_TO_QSCRIPTVALUE_GETTER_ALWAYS(type, EntityTypes::getEntityTypeName(_type));
     }
-    if (!psuedoPropertyFlagsActive || psueudoPropertyFlags.test(EntityPsuedoPropertyFlag::Created)) {
```
Comment (SamGondelman) [inline libraries/entities/src/EntityItemProperties.cpp]: `created` is not a pseudo property, it needs to be written all the time

File: libraries/entities/src/GrabPropertyGroup.cpp [hunk 8]
```diff
@@ -106,6 +108,9 @@ void GrabPropertyGroup::listChangedProperties(QList<QString>& out) {
     if (equippableChanged()) {
         out << "grab-equippable";
     }
+    if (grabDelegateToParentChanged()) {
+        out << "grab-grabDelegateToParent";
+    }
```
Comment (AndrewMeadows) [inline libraries/entities/src/GrabPropertyGroup.cpp]: Not a blocker but:  this naming pattern:  `foo-fooFuBar` makes me cry.  I wish it were `foo-fuBar`. :cry:

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16269/HighFidelity-Beta-PR14657-b5b7574.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16269/HighFidelity-Beta-Interface-PR14657-b5b7574.exe

**Comment by qlimnavarro** [comment]
Currently testing

**Comment by qlimnavarro** [comment]
Fail
Originally defect still occurs with the following steps:
1. Enter script into Console and run it
2. Hat appears and passes test as expected
3. Re-Enter script 
4. hat appears and immediately disappears

[hifi-log.txt](https://github.com/highfidelity/hifi/files/2733491/hifi-log.txt)

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16282/HighFidelity-Beta-PR14657-d57e3b5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16282/HighFidelity-Beta-PR14657-d57e3b5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16282/HighFidelity-Beta-Interface-PR14657-d57e3b5.dmg

**Comment by qlimnavarro** [comment]
Currently Testing when built

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16282/HighFidelity-Beta-PR14657-d57e3b5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16282/HighFidelity-Beta-Interface-PR14657-d57e3b5.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16295/HighFidelity-Beta-PR14657-fa66ed7.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16295/HighFidelity-Beta-PR14657-fa66ed7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16295/HighFidelity-Beta-Interface-PR14657-fa66ed7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16295/HighFidelity-Beta-PR14657-fa66ed7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16295/HighFidelity-Beta-Interface-PR14657-fa66ed7.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16299/HighFidelity-Beta-PR14657-4210d4a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16299/HighFidelity-Beta-PR14657-4210d4a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16299/HighFidelity-Beta-Interface-PR14657-4210d4a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16299/HighFidelity-Beta-PR14657-4210d4a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16299/HighFidelity-Beta-Interface-PR14657-4210d4a.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16305/HighFidelity-Beta-PR14657-791257b.apk

**Comment by qlimnavarro** [comment]
Pass
Wearables no longer disappear

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16305/HighFidelity-Beta-PR14657-791257b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16305/HighFidelity-Beta-Interface-PR14657-791257b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16305/HighFidelity-Beta-PR14657-791257b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16305/HighFidelity-Beta-Interface-PR14657-791257b.exe

**Comment by SamGondelman** [comment]
@JeffClinton @qlimnavarro just merged from master to resolve a conflict, still no retest needed

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16306/HighFidelity-Beta-PR14657-db3187e.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16306/HighFidelity-Beta-PR14657-db3187e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16306/HighFidelity-Beta-Interface-PR14657-db3187e.exe

**Pull request merged**